posts/lambda-execution-context-demystified.md

```diff
@@ -0,0 +1,128 @@
+---
+authors:
+- Raphael Brugier
+tags:
+- AWS
+- AWS Lambda
+- Java
+date: 2018-10-17T18:05:14.000Z
+title: "AWS Lambda Execution context in Java demystified"
+image: image: https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2017/12/te-le-chargement.png
+---
+
+
+In a [previous article](https://blog.ippon.tech/serverless-architectures/), we described how Serverless architectures enable scalability and can reduce costs. As of now, the most popular cloud provider for Serverless functions is still AWS with its Lambda service.
+
+In this article, I will attempt to demystify the _Execution Context_ of a Lambda invocation and how you can take advantage of it. More specifically for Lambdas written in Java, but this can apply to any language.
+
+# Lambda Execution Context
+
+In the [AWS Lambda documentation](https://docs.aws.amazon.com/lambda/latest/dg/running-lambda-code.html), AWS describes an Execution Context as a “temporary runtime environment that initializes any external dependencies of your Lambda”.
+
+The _Execution Context_ is the invisible stack that AWS creates for your Lambda in order to execute the function. AWS maintains the _Context_ ready to accept new invocations of the function for an unknown amount of time[^1] and then delete it to free up some resource. Between each invocation, AWS freezes and unfreezes the Context.
+
+For a Lambda written in Java, this is how I like to represent the different layers of the _Execution Context_:
+
+![](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2018/10/lambda-java-execution-context.png)
+
+- An EC2 instance runs the entire stack. It’s not possible to physically access the instance, and AWS does not share information about this layer
+- Container: a container is started to run each Lambda in isolation. Multiple containers run in the same instance
+- JVM: For Lambdas written in Java, a Java Virtual Machine is started. AWS uses OpenJDK 8
+- An instance of the JavaHandler class where the function is defined
+- The function itself
+
+Between each invocation of the `handleRequest` method, the same _Execution Context_ is reused. This, at least until the Lambda function does not receive requests anymore and AWS destroys the _Context_ to reclaim the resources.
+
+
+# Why it matters
+As you can imagine, starting-up the _Execution Context_ takes more time than just invoking the function. 
```

Comment:
"takes more time than just invoking the function" -> not necessarily true / depends on the function

Reply:
ah yes, absolutely. Especially now that a Lambda can run up to 15 min :) 
```suggestion
As you can imagine, starting-up the _Execution Context_ can take some time but is inevitable before running the function for the first time.
```